crypto/provider_core.c

```diff
@@ -1022,6 +1052,33 @@ void ossl_provider_unquery_operation(const OSSL_PROVIDER *prov,
         prov->unquery_operation(prov->provctx, operation_id, algs);
 }
 
+int ossl_provider_clear_all_operation_bits(OSSL_LIB_CTX *libctx)
+{
+    struct provider_store_st *store;
+    OSSL_PROVIDER *provider;
+    int i, num, res = 1;
+
+    if ((store = get_provider_store(libctx)) != NULL) {
+        if (!CRYPTO_THREAD_write_lock(store->lock))
```

Comment:
Is a read lock on the store sufficient since you are not changing the store itself?

Reply:
I'm pretty sure it's required.  Something needs to prevent the provider list being changed whilst traversing it.

Reply:
A read lock is sufficient to prevent that

Reply:
The providers _in the_ list, the `operation_bits` can be queried while this code is clearing them and flushing  algorithms.  This could lead to some weird things: process A checks the bits, process B flushes, A does something with them.  When the only operations are set and query, this is safe.

Thinking about it more, I'm not sure the write lock will prevent trouble since the set and query don't lock the store.  I'll change to a read lock and see what explodes :)

Do you remember why the `opbits_lock` is used instead of the store's main lock?


BTW: lines 239-245 in the ossl_provider_find() function use a write lock for what seems like a read too:

https://github.com/openssl/openssl/blob/1127754e4877b2a4bd53112de115041d1952fa12/crypto/provider_core.c#L219-L249

Reply:
Hmm. Ok.